File: server/app/schemas/db/resume.py
from __future__ import annotations
from typing import List, Union
from pydantic import BaseModel


class Resume(BaseModel):
    class ResumeSection(BaseModel):
        title: str
        content: List[Union[str, Resume.ResumeSection]]

    _id: str
    root: List[ResumeSection]

File: frontend/src/types/resume.ts
import { z } from "zod";

type _ResumeSection = {
  title: string;
  content: Array<string | ResumeSection>;
};

const ResumeSectionSchema: z.ZodType<_ResumeSection> = z.lazy(() =>
  z.object({
    title: z.string(),
    content: z.array(z.union([z.string(), ResumeSectionSchema])),
  }),
);

export const ResumeSchema = z.object({
  root: z.array(ResumeSectionSchema),
});

export type ResumeSection = z.infer<typeof ResumeSectionSchema>;
export type Resume = z.infer<typeof ResumeSchema>;

export type ParsedResumeSection = ResumeSection & {
  startDate?: Date;
  endDate?: Date;
};
